Dynamic socketpool

i110 opened the pull request:
Make it possible to use keep-alive connections for reproxy or mruby's `http_request` method

- [x] fix overall design and interfaces
- [x] ~~use O(1) way to lookup existing target~~ -> fix later when actual problems emerge
- [x] ~~add configuration nobs and parameters (capacity, timeout, etc)~~
- [x] address race condition issues around pool->targets modification
- [x] `location_rewrite_url` ?
- [x] add tests

kazuho commented on lib/common/socketpool.c:
```diff
@@ -229,6 +239,42 @@ void h2o_socketpool_init_target_by_hostport(h2o_socketpool_target_t *target, h2o
+    return NULL;
+}
+
+h2o_socketpool_target_t *h2o_socketpool_get_or_add_target(h2o_socketpool_t *pool, h2o_iovec_t host, uint16_t port, int is_ssl, h2o_url_t *url)
```
Current design seems to expose a function that searches for or adds a target (i.e. `get_or_add_target`) and passes the address of the target to the function that actually connects to the server.

The approach seem to have two issues:
* once allocated, `pool->targets` becomes unexpandable (or shrinkable), unless the mutex is locked from before the call to `get_or_add_target` to after the call to `connect_to_target`. Otherwise, we would have a race condition.
* a target entry would be created for a server that always sends `connection: close`, which is IMO a waste

So you might want to consider the following:
* do not expose `get_or_add_target`
* accept the (host, port, is_ssl) triplet as an argument to `connect_to_target`
 * the function checks if a matching target exists, and if a target with a pooled socket found, returns the socket
 * otherwise, the socket pool connects to a new socket (note that you do not need to ensure that a matching target exists)
* in `h2o_socketpool_return`, look for a matching target, and if _not_ found, create a new target. Add to the target the socket being returned. You might need to pass the triplet as arguments to the function.

By using this approach, all locking operations can be kept within the socketpool.c. Also, `pool->targets` becomes not only expandable but also shrinkable (i.e. targets without any poold sockets can be garbage-collected).

kazuho commented on include/h2o.h:
```diff
@@ -415,6 +415,9 @@ struct st_h2o_globalconf_t {
         unsigned emit_via_header : 1;
+
+        int use_dynamic_socketpool : 1;
+        h2o_socketpool_t *dynamic_socketpool;
```
I believe that there is no reason to delay the instantiation of the pool. Let's have it initialized in src/main.c.

i110 commented:
@kazuho I addressed the issues you pointed, and now this PR contains certain design changes around socketpool and http1client.

- two types of socketpool
  - static socketpool -  it's initialized with `h2o_url_t`'s, and converts those into h2o_socketpool_target_t internally, and no more changes of targets are permitted.
  - dynamic socketpool - it can be added new targets after initialization
- http1client always (even when keepalive is disabled) connect to the peer using socketpool. if the `socketpool` argument is NULL, it uses default socketpool. By doing this, the implementation become very simple
- the `location_rewrite_url` argument of `h2o_socketpool_connect` function is removed because (as far as I checked) it always equals the origin of selected target.

i110 commented:
cc: @deweerdt @zlm2012

kazuho commented on include/h2o/socketpool.h:
```diff
@@ -95,29 +101,15 @@ typedef struct st_h2o_socketpool_t {
+typedef void (*h2o_socketpool_connect_cb)(h2o_socket_t *sock, const char *errstr, void *data, h2o_url_t *url);
 /**
- * initializes a socket loop
+ * initializes a static socket pool
```
How about using words: _specific_ and _global_ instead of (_static_ and _dynamic_)?

kazuho commented on lib/common/socketpool.c:
```diff
@@ -37,6 +37,7 @@ struct pool_entry_t {
     h2o_socket_export_t sockinfo;
     h2o_socketpool_target_t *target;
     h2o_linklist_t link;
```
I would appreciate it if you could rename `link` to `all_link` so that the roles of the two links (the other is `target_link`) become clearer.

kazuho commented on lib/common/socketpool.c:
```diff
@@ -47,13 +48,18 @@ struct st_h2o_socketpool_connect_request_t {
+            int *tried;
+            size_t try_count;
+        } lb;
+    };
```
Can't this be changed to something like:
```
h2o_multithread_receiver_t *getaddr_receiver;
size_t selected_target;
struct {
    int *tried;
    size_t try_count;
} lb;
```

What I am suggesting here is that you only need to retain the selected index of that designates the target within `pool->targets`. I believe that not using unions would reduce the impact on the code-base.

Note that the proposed struct removes `lb.targets`. I believe that it is possible; please see 5780763.

zlm2012 commented:
@i110 

> the `location_rewrite_url` argument of `h2o_socketpool_connect` function is removed because (as far as I checked) it always equals the origin of selected target.

Sure. That is the exact meaning of that argument. I just didn't come out better naming of that argument and it was meant that `location_rewrite_url` should be rewritten if it appeared in the `Location` header responded from upstream.

And sorry for the delay, just come back from vacation :-)

i110 commented:
@kazuho addressed issues you pointed.

kazuho commented on lib/common/socketpool.c:
```diff
@@ -406,30 +443,49 @@ void h2o_socketpool_connect(h2o_socketpool_connect_request_t **_req, h2o_socketp
-        entry = H2O_STRUCT_FROM_MEMBER(struct pool_entry_t, link, pool->_shared.sockets.next);
-        h2o_linklist_unlink(&entry->link);
+
+    /* TODO is this needed in this critical section? */
```
No. I think that you should move the call to `lookup_target` outside of the critical section.

i110 commented:
@kazuho done!

kazuho commented:
@i110 With my four commits (listed above) I think the PR is ready for merge. Please let me know what you think.

i110 commented:
@kazuho the code gets cleaner, thanks!
